Update dynamo sum dtype handling to match eager

ekamiti opened the pull request:
The current behaviour for dynamo is to set the dtype to torch.int64 for integral types if the dtype is not specified explicitly which results in mismatched behaviour as compared to eager mode. In eager mode the semantics are:
- If both out is specified and dtype is specified then they have to match
- If dtype is not specified but out is specified then the dtype is set to match the out dtype
- If neither dtype nor out is set then the dtype is set to kLong if it is a bool or an integral type

Fixes #100698

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @msaroufim @aakhundov @bertmaher

ekamiti committed:
Update dynamo sum dtype handling to match eager

The current behaviour for dynamo is to set the dtype to torch.int64 for
integral types if the dtype is not specified explicitly which results in
mismatched behaviour as compared to eager mode. In eager mode the semantics are:
- If both out is specified and dtype is specified then they have to match
- If dtype is not specified but out is specified then the dtype is set to match the out dtype
- If neither dtype nor out is set then the dtype is set to kLong if it is a bool or an integral type

janeyx99 commented:
@yanboliang assigned you as reviewer as you recently touched these files, but feel free to reassign

yanboliang commented on test/dynamo/test_repros.py:
```diff
@@ -3284,6 +3284,72 @@ def fn(x):
         ref = opt_fn(x)
         self.assertEqual(ref, res)
 
+    def test_sum_dtype(self):
```
Actually this is not a Dynamo bug, it's a decomposition bug, it's better we can move the unit test to ```torch/testing/_internal/common_methods_invocations.py```, you can just add a similar item like this:
https://github.com/pytorch/pytorch/blob/dfa64fddeb6b58c60a59096202b192ea09f6ed40/torch/testing/_internal/common_methods_invocations.py#L20214-L20219

ekamiti replied:
Thanks for the review. I see that above this there is: https://github.com/pytorch/pytorch/blob/dfa64fddeb6b58c60a59096202b192ea09f6ed40/torch/testing/_internal/common_methods_invocations.py#L20209-L20213

I just want to confirm that you'd like me to add a `PythonRefInfo` in addition to this `ReductionPythonRefInfo`, is that the case?

yanboliang replied:
That's a good question, I'm not pretty sure. @ngimel can you suggest which is the right way?

ekamiti replied:
If I understand the test correctly, the ops test that maybe could have caught this issue is this one: 
https://github.com/pytorch/pytorch/blob/dfa64fddeb6b58c60a59096202b192ea09f6ed40/test/test_ops.py#L786-L802

But I don't think it actually tests this behavior at all since it always provides a dtype in all the cases, and the bug fixed by this PR requires that the dtype not be provided. It also doesn't exercise the case where the provided dtype is an integral type which would be exercised in case 4 but that only runs if the dtype is a floating point type.

yanboliang reviewed (commented):
the change looks good to me, had a suggestion to move unit tests to the right place.

ekamiti committed:
Move Out Dtype test for reduction to correct file

Moved the test from test_repros.py to test/test_ops.py and also
fixed the same error that was affecting sum in prod.

ngimel commented on test/test_ops.py:
```diff
@@ -981,6 +992,37 @@ def _case_four_transform(t):
+                else:
+                    out = op_to_test(inputs, *args, **kwargs)
+                self.assertFalse(expectFail)
+            except RuntimeError:
```
don't catch just runtime error, make sure that the error is what you expect

ekamiti replied:
Thank you for the review! `RuntimeError` is the error thrown, it can be found here:

https://github.com/pytorch/pytorch/blob/d89dd05e4d25c25d5c95be113a45a6ee61f5d85e/torch/_refs/__init__.py#L2059-L2066

ngimel replied:
I understand that type of the error is "RuntimeError", but you should also check error message, otherwise many errors including those you don't expect will pass.

ekamiti replied:
I have added an explicit check for the message.

ngimel replied:
Cool, feel free to land with pytorchbot command when CI looks good.

ngimel reviewed (approved):
This lgtm, test is fine

ekamiti committed:
Make sure error message for exception is correct

ekamiti committed:
Fix lintrunner failure

ekamiti commented:
@pytorchbot merge

ekamiti commented:
@pytorchbot label "release notes: dynamo"

ekamiti commented:
@pytorchbot merge

ekamiti commented:
@ngimel @yanboliang So this test is failing because of this test case:

https://github.com/pytorch/pytorch/blob/1eb762c919ba3bd1809671050cd96d31d2b85757/test/test_ops.py#L946-L953

Which does the type of operation below and expects it to fail. Seems like a reasonable failure case though the behavior that it has right now mirrors that of eager mode so not really sure whether to skip this test for the affected operators or try to prevent type promotion from float to int. Any thoughts would be appreciated.

```
>>> a = torch.tensor([5, 1, 6, 5, 6], dtype=torch.float32)
>>> out = torch.empty(0, dtype=torch.int64)
>>> def wrap_sum(*args, **kwargs):
...     return torch.sum(*args, **kwargs)
...
>>> dynamo_sum = torch.compile(wrap_sum)
>>> torch.sum(a, 0, out=out)
tensor(23)
>>> out = torch.empty(0, dtype=torch.int64)
>>> dynamo_sum(a, 0, out=out)
tensor(23)
```

ngimel commented:
Yeah longer term it would be good to fix eager behavior wrt to unsafe casts, but for now it's ok to skip the test (you can add the skip in the corresponding OpInfos in common_methods_invocations.py). Thank you!

ekamiti committed:
Disable tests that don't pass test_out with the
fix as they don't test the out operator correctly
for sum, prod, cumsum, and cumprod

ekamiti commented:
@ngimel 
> Yeah longer term it would be good to fix eager behavior wrt to unsafe casts, but for now it's ok to skip the test (you can add the skip in the corresponding OpInfos in common_methods_invocations.py). Thank you!

On a related note, I realized there are some existing tests that only run for eager mode operations and not for the _refs.

https://github.com/pytorch/pytorch/blob/93b0410eef57ebf038c12ed2fa1d4018a24096b7/torch/testing/_internal/common_methods_invocations.py#L20599-L20611

When I for example change the set of reduction_ops to include the refs_ some of the test cases fail. Seems like there may be more cases of mismatched functionality between eager and compiled operators. Is this lack of coverage intentional or should a bug be filled for it if there isn't one?

ekamiti commented:
@pytorchbot merge

ekamiti committed:
Set device type when creating out tensor

ekamiti commented:
@pytorchbot merge

huydhn commented:
@pytorchbot rebase

huydhn commented:
@pytorchbot merge

huydhn commented:
@pytorchbot merge

huydhn commented:
@pytorchbot rebase -b main

ekamiti commented:
@pytorchbot rebase -b main

ekamiti commented:
@huydhn I see my merge didn't succeed because of some failures though they don't look related to my changes at all as far as I can tell from the logs and the fact that runs of the same tests under different configurations seem to have passed and the changes made in this PR should not be dependent of any device specific configurations. What's the best way to proceed here? Thanks for all your help so far!

huydhn commented:
@pytorchbot merge

huydhn commented:
@pytorchbot merge -f 'Unrelated trunk and unstable failures'
